src/QSymbolicsBase/linalg.jl

```diff
@@ -89,6 +89,160 @@ function Base.show(io::IO, x::SDagger)
+tr(x::STensorOperator) = (*)((tr(i) for i in arguments(x))...)
+Base.hash(x::STrace, h::UInt) = hash((head(x), arguments(x)), h)
+Base.isequal(x::STrace, y::STrace) = isequal(x.op, y.op)
+
+"""Partial trace over system i of a composite quantum system
+
+```jldoctest
+julia> @op 𝒪 SpinBasis(1//2)⊗SpinBasis(1//2);
+
+julia> op = ptrace(𝒪, 1)
+tr1(𝒪)
+
+julia> QuantumSymbolics.basis(op)
+Spin(1/2)
+
+julia> @op A; @op B;
+
+julia> ptrace(A⊗B, 1)
+(tr(A))B
+
+julia> @ket k; @bra b;
+
+julia> pure_state = A ⊗ (k*b)
+(A⊗|k⟩⟨b|)
+
+julia> ptrace(pure_state, 1)
+(tr(A))|k⟩⟨b|
+
+julia> ptrace(pure_state, 2)
+(⟨b||k⟩)A
+
+julia> mixed_state = (A⊗(k*b)) + ((k*b)⊗B)
+((A⊗|k⟩⟨b|)+(|k⟩⟨b|⊗B))
+
+julia> ptrace(mixed_state, 1)
+((0 + ⟨b||k⟩)B+(tr(A))|k⟩⟨b|)
+
+julia> ptrace(mixed_state, 2)
+((0 + ⟨b||k⟩)A+(tr(B))|k⟩⟨b|)
+```
+"""
+@withmetadata struct SPartialTrace <: Symbolic{AbstractOperator}
+    obj
+    sys::Int
+end
+isexpr(::SPartialTrace) = true
+iscall(::SPartialTrace) = true
+arguments(x::SPartialTrace) = [x.obj, x.sys]
+operation(x::SPartialTrace) = ptrace
+head(x::SPartialTrace) = :ptrace
+children(x::SPartialTrace) = [:ptrace, x.obj, x.sys]
+function basis(x::SPartialTrace)
+    obj_bases = collect(basis(x.obj).bases)
+    new_bases = deleteat!(copy(obj_bases), x.sys)
+    tensor(new_bases...)
+end
+Base.show(io::IO, x::SPartialTrace) = print(io, "tr$(x.sys)($(x.obj))")
+function ptrace(x::Symbolic{AbstractOperator}, s) 
+    ex = isexpr(x) ? qexpand(x) : x
+    if isa(ex, typeof(x))
+        if isa(basis(x), CompositeBasis)
+            SPartialTrace(x, s)
+        elseif s==1
+            tr(x)
+        else
+            throw(ArgumentError("cannot take partial trace of a single quantum system"))
+        end
+    else
+        ptrace(ex, s)
+    end
+end
+function ptrace(x::SAddOperator, s)
+    add_terms = []
+    if isa(basis(x), CompositeBasis)
+        for i in arguments(x)
+            if isexpr(i)
+                if isa(i, SScaledOperator) && operation(i.obj) === ⊗
+                    prod_terms = arguments(i.obj)
+                    coeff = i.coeff
+                elseif operation(i) === ⊗
+                    prod_terms = arguments(i)
+                    coeff = 1
+                else
+                    return SPartialTrace(x,s)
+                end
+            else
+                return SPartialTrace(x,s)
+            end
+            if any(j -> isa(basis(j), CompositeBasis), prod_terms)
+                return SPartialTrace(x,s)
+            else
+                sys_op = coeff*prod_terms[s]
+                new_terms = deleteat!(copy(prod_terms), s)
+                trace = tr(sys_op)
+                isone(length(new_terms)) ? push!(add_terms, trace*first(new_terms)) : push!(add_terms, trace*(⊗)(new_terms...))
+            end  
+        end
+        (+)(add_terms...)
+    elseif s==1
+        tr(x)
+    else
+        throw(ArgumentError("cannot take partial trace of a single quantum system"))
+    end
+end
```

Comment:
I should explain my thinking behind this complicated method, as it would be good for me to have my thoughts organized in writing and it would be helpful for review.

 It is intended to take the partial trace of a summation of operators. First, we must check if the basis of the summation is a composite basis, and if not, then the partial trace must be over the first system (in that case we would just take a regular trace of the sum). If neither of these conditions are true, then we throw an argument error, as then we would be taking the partial trace of a single quantum system.

If the basis of the summation is a composite basis, then we iterate through the arguments of the sum. Here is where it gets complicated. A term in the arguments could be (1) a tensor product of operators with non-composite bases, (2) a tensor product of operators with composite bases, (3) an operator with composite basis (could be `SMulOperator` or just a single operator, or (4) a scaled version of (1-3). That is why we have that first if-else statement in the for loop.

Then, we have the `if any(j ->...` statement to deal with the case where we have a tensor product of operators with composite bases, in which case we can't take the trace of any of the operators, so we just return `SPartialTrace`. If that condition is not satisfied, i.e., we have a summation of tensor products of operators, each with non-composite bases, then we proceed with tracing over the specified system.

Hopefully this helps. I'm happy to clarify anything.

Reply:
Ah yeah, maybe we could use rewriters from Symbolics for this. Although they can get messy too, especially when you have multiple predicates involved. When I was making this, I was mostly focused on capturing all of the different cases and making it easy to follow (even if it is indeed long). I'll think about this more but for now I say we leave it as is.